radicle-node/src/worker/fetch.rs

```diff
@@ -0,0 +1,333 @@
+    staging: Storage,
+    /// The original [`Storage`] we are finalising changes into.
+    production: &'a Storage,
+    _tmp: tempfile::TempDir,
+}
+
+/// Indicates whether the innner [`Repository`] is being cloned into
+/// or fetched into.
+pub enum StagedRepository {
+    Cloning(Repository),
+    Fetching(Repository),
+}
+
+impl Deref for StagedRepository {
+    type Target = Repository;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            StagedRepository::Cloning(repo) => repo,
+            StagedRepository::Fetching(repo) => repo,
+        }
+    }
+}
+
+/// The second staging area to perform a fetch from a remote.
+///
+/// The [`StagingRefs::refpsecs`] generated are to fetch any follow-up
+/// references after the fetch on [`StagingRad`]. This may be all the
+/// delegate's references in the case of cloning the new repository,
+/// or it could be fetching the latest updates in the case of fetching
+/// an existing repository.
+///
+/// It is then expected to finalise the process by transferring the
+/// fetched references into the production storage, via
+/// [`StagingRefs::transfer`].
+pub struct StagingRefs<'a> {
+    /// The inner [`Repository`] for staging fetches into.
+    pub(super) repo: StagedRepository,
+    /// The staging [`Storage`] for staging the `repo`.
+    staging: Storage,
+    /// The original [`Storage`] we are finalising changes into.
+    production: &'a Storage,
+    /// The remotes that the fetch is being performed for. These are
+    /// discovered after performing the fetch for [`StagingRad`].
+    remotes: NonEmpty<RemoteId>,
+    _tmp: tempfile::TempDir,
+}
+
+enum VerifiedRemote {
+    Failed {
+        reason: String,
+    },
+    Success {
+        // XXX: unused but we want to ensure that we verify the identity
+        _doc: Doc<Verified>,
+        remote: Remote<Verified>,
+    },
+}
+
+impl<'a> StagingRad<'a> {
+    /// Construct a [`StagingRad`] which sets up its
+    /// [`StagedRepository`] in a new, temporary directory.
+    pub fn new(production: &'a Storage, rid: Id) -> Result<Self, error::Create> {
+        let tmp = tempfile::TempDir::new()?;
+        log::debug!(target: "fetch", "Staging fetch in {:?}", tmp.path());
+        let staging = Storage::open(tmp.path())?;
+        let repo = Self::repository(&staging, production, rid)?;
+        Ok(Self {
+            repo,
+            staging,
+            production,
+            _tmp: tmp,
+        })
+    }
+
+    /// Return the fetch refspecs for fetching the necessary `rad`
+    /// references.
+    pub fn refspecs(
+        &self,
+        namespaces: Namespaces,
+    ) -> Vec<Refspec<git::PatternString, git::PatternString>> {
+        let id = git::PatternString::from(IDENTITY_BRANCH.clone().into_refstring());
+        match self.repo {
+            StagedRepository::Cloning(_) => Refspec {
+                src: id.clone(),
+                dst: id,
+                force: true,
```

Comment:
Why would `force` be true if this is cloning?  My belief being there's nothing to force write over, since something is being cloned.

Reply:
Ya, you're right. If `force: false` fails then something went wrong so it should be set to `false` in the cloning case :+1: